ipn/ipnlocal/local.go

```diff
@@ -7864,6 +7886,99 @@ func suggestExitNode(report *netcheck.Report, nb *nodeBackend, prevSuggestion ta
 	return res, nil
 }
 
+var ErrNoNetMap = errors.New("no network map, try again later")
+
+// suggestExitNodeUsingTrafficSteering uses traffic steering priority scores to
+// pick one of the best exit nodes. These priorities are provided by Control in
+// the node’s [tailcfg.Location]. To be eligible for consideration, the node
+// must have NodeAttrSuggestExitNode in its CapMap.
+func suggestExitNodeUsingTrafficSteering(nb *nodeBackend, prev tailcfg.StableNodeID, allowed set.Set[tailcfg.StableNodeID]) (apitype.ExitNodeSuggestionResponse, error) {
+	nm := nb.NetMap()
+	if nm == nil {
+		return apitype.ExitNodeSuggestionResponse{}, ErrNoNetMap
+	}
+
+	if !nb.SelfHasCap(tailcfg.NodeAttrTrafficSteering) {
+		panic("missing traffic-steering capability")
+	}
+
+	peers := nm.Peers
+	nodes := make([]tailcfg.NodeView, 0, len(peers))
+
+collect:
+	for _, p := range peers {
+		switch {
+		case !p.Valid():
+		case allowed != nil && !allowed.Contains(p.StableID()):
+		case !p.CapMap().Contains(tailcfg.NodeAttrSuggestExitNode):
+		case !tsaddr.ContainsExitRoutes(p.AllowedIPs()):
+		case p.StableID() == prev:
+			// Prevent flapping: since prev is a valid suggestion,
+			// force prev to be the only valid pick.
+			nodes = []tailcfg.NodeView{p}
+			break collect
+		default:
+			nodes = append(nodes, p)
+		}
+	}
+
+	var pick tailcfg.NodeView
+
+	scores := make(map[tailcfg.NodeID]int, len(nodes))
+	score := func(n tailcfg.NodeView) int {
```

Comment:
This feels like it would be a good fit for `lazyGValueMap` from `tailnet_mapper.go` but all that's in OSS is `lazy.GValue`, for single entries. This isn't too complicated, it's fine as-is.

Reply:
Good to know.